plugins/tidy-links.js

```diff
@@ -24,59 +24,103 @@ tidyLinks.STROKE_STYLE = { // https://leafletjs.com/reference-1.4.0.html#polylin
 
 var map;
 
-tidyLinks.getLocations = function (limit) {
+tidyLinks.getLocations = function () {
+  var markers = [];
+  var polygons = [];
+  if (plugin.drawTools) {
+    markers = window.plugin.drawTools.getMarkers().map(function(layer) { return layer.getLatLng(); });
+    var geoPolygons = window.plugin.drawTools.getPolygons();
+    for (var i = 0; i < geoPolygons.length; i++) {
+      if (geoPolygons[i]._rings && geoPolygons[i]._rings.length)
+        polygons.push(geoPolygons[i]._rings.flat());
+    }
+  }
+
+  // fallback to map bounds if no drawn polygon (or no drawtools)
+  if (!polygons.length) {
+    var bounds = map.getPixelBounds();
+    var origin = map.getPixelOrigin();
+    polygons.push([
+      bounds.getBottomLeft().subtract(origin),
+      bounds.getBottomRight().subtract(origin),
+      bounds.getTopRight().subtract(origin),
+      bounds.getTopLeft().subtract(origin),
+    ]);
```

Comment:
I do not like this approach.
Tidy links should not count draw-tools polygons.
And tidy links should not even call pnpoly directly.

Things must be much simpler, and more high-level: pass point - get back boolean.

Reply:
Tidy/Fly links need to know about draw polygons : 
![image](https://user-images.githubusercontent.com/64744459/110257699-57882580-7f9f-11eb-806f-519b4d3a7770.png)

Reply:
Not sure that it's really useful.
If you like this then ok.

But I still propose to invent better API.
E.g. special draw-tools function `getFilters`.
And remember that we have to react on drawings change.

Reply:
I'd like to see all low-level details incapsulated (inside of draw-tools).
And all plugins access just high-level api.

- `locationFilters:Array` of filters, with such interface:
  + `filter(portal: CircleMarker|LatLng)`.
- Also we need "changed" event to be able to react on selection changes.

And that's all.
Client plugins should not care about points format, bounding polygons, projections, exclusion markers and lines.

Reply:
It's useful, for instance you want to max AP but a long link cuts the area in two parts.
And, that's current fly links feature.

This PR is about fixing Fly link when drawtools isn't available and add the same feature to tidy links
We can always imagine better API, but it should match the use cases at least.

Here, we'd like a "selection" API to return the sets of portal we are in, their colors for instance, in order to run the processing on each set.

Reply:
> And, that's current fly links feature.

Ok

> We can always imagine better API, but it should match the use cases at least.

Better API - simpler api. But sure we should cover most use cases.

> we'd like a "selection" API to return the sets of portal we are in, their colors for instance, 

That would be redundant overcomplication.

Reply:
> This PR is about fixing Fly link when drawtools isn't available

That could be merged in it's initial form.

> and add the same feature to tidy links

But this demands more work.

Reply:
> But this demands more work.

but it doesn't need either a new API. Here common parts are barely factorized to avoid algorithm duplication. Fly/tidy are coupled to drawtools (forcing links in fly links for instance) and that is fine.

pnpoly can be change later (preserving the signature once released)

Select API is something to be discussed in parallel.

> That would be redundant overcomplication.

why redundant ? 
```
getSelectRegions(): id[];
isSelected(portal/latlng,id?): boolean;
getSelectedPortals(id?): guids/portals[];
```
this would lead to code like 
```
for (id in getSelectRegions()) {
  doSomethingWith(getSelectedPortal(id));
}
```
with something like for instance
```
addFilter(filter: (latlng/portal)=>boolean,id?): id;
```

but this is another discussion imo

Reply:
> but it doesn't need either a new API

I have severe objections both against massive copy-paste, and early unmature api decisions.

> why redundant ?

I mean redundant in API, but could be implemented in draw-tools interface instead.

> getSelectRegions(): id[]; // etc..

That has exactly the same functionality as proposed above https://github.com/IITC-CE/ingress-intel-total-conversion/pull/544#discussion_r762807857
No?

Reply:
> I have severe objections both against massive copy-paste, and early unmature api decisions.

sure. but here, there is no API change right ? (and copy/paste was removed)
I especially don't want to add filter/select api here because I find it not mature enough

> That has exactly the same functionality as proposed above

yes, this was an illustration (with different naming) of what was saying about : (wrt to redundancy)
> return the sets of portal _[by id/index]_ we are in, their colors for instance

Reply:
> there is no API change right ?

window.pnpoly, window.pnpolyLL

> because I find it not mature enough

It will get more mature as we develop it. In this PR, or in another.

Reply:
point in polygon is not restricted to filter/select, that could even extend L.Polygon.
that's why it's exposed regardless of the selection api you want to build